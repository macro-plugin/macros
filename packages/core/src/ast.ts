import type { ArrayExpression, ArrayPattern, ArrowFunctionExpression, AssignmentExpression, AssignmentPattern, AssignmentPatternProperty, AssignmentProperty, AwaitExpression, BigIntLiteral, BinaryExpression, BlockStatement, BooleanLiteral, BreakStatement, CallExpression, CatchClause, ClassDeclaration, ClassExpression, ClassMethod, ClassProperty, ComputedPropName, ConditionalExpression, Constructor, ContinueStatement, DebuggerStatement, Decorator, DoWhileStatement, EmptyStatement, ExportAllDeclaration, ExportDeclaration, ExportDefaultDeclaration, ExportDefaultExpression, ExportDefaultSpecifier, ExportNamedDeclaration, ExportNamespaceSpecifier, ExpressionStatement, ForInStatement, ForOfStatement, ForStatement, FunctionDeclaration, FunctionExpression, GetterProperty, Identifier, IfStatement, Import, ImportDeclaration, ImportDefaultSpecifier, ImportNamespaceSpecifier, Invalid, JSXAttribute, JSXClosingElement, JSXClosingFragment, JSXElement, JSXEmptyExpression, JSXExpressionContainer, JSXFragment, JSXMemberExpression, JSXNamespacedName, JSXOpeningElement, JSXOpeningFragment, JSXSpreadChild, JSXText, KeyValuePatternProperty, KeyValueProperty, LabeledStatement, MemberExpression, MetaProperty, MethodProperty, Module, NamedExportSpecifier, NamedImportSpecifier, NewExpression, NullLiteral, NumericLiteral, ObjectExpression, ObjectPattern, OptionalChainingCall, OptionalChainingExpression, Param, ParenthesisExpression, PrivateMethod, PrivateName, PrivateProperty, RegExpLiteral, RestElement, ReturnStatement, Script, SequenceExpression, SetterProperty, SpreadElement, StaticBlock, StringLiteral, Super, SuperPropExpression, SwitchCase, SwitchStatement, TaggedTemplateExpression, TemplateElement, TemplateLiteral, ThisExpression, ThrowStatement, TryStatement, TsArrayType, TsAsExpression, TsCallSignatureDeclaration, TsConditionalType, TsConstAssertion, TsConstructSignatureDeclaration, TsConstructorType, TsEnumDeclaration, TsEnumMember, TsExportAssignment, TsExpressionWithTypeArguments, TsExternalModuleReference, TsFunctionType, TsGetterSignature, TsImportEqualsDeclaration, TsImportType, TsIndexSignature, TsIndexedAccessType, TsInferType, TsInstantiation, TsInterfaceBody, TsInterfaceDeclaration, TsIntersectionType, TsKeywordType, TsLiteralType, TsMappedType, TsMethodSignature, TsModuleBlock, TsModuleDeclaration, TsNamespaceDeclaration, TsNamespaceExportDeclaration, TsNonNullExpression, TsOptionalType, TsParameterProperty, TsParenthesizedType, TsPropertySignature, TsQualifiedName, TsRestType, TsSetterSignature, TsTemplateLiteralType, TsThisType, TsTupleElement, TsTupleType, TsTypeAliasDeclaration, TsTypeAnnotation, TsTypeAssertion, TsTypeLiteral, TsTypeOperator, TsTypeParameter, TsTypeParameterDeclaration, TsTypeParameterInstantiation, TsTypePredicate, TsTypeQuery, TsTypeReference, TsUnionType, UnaryExpression, UpdateExpression, VariableDeclaration, VariableDeclarator, WhileStatement, WithStatement, YieldExpression } from "@swc/core"

export interface AST {
 "Identifier": Identifier
 "StringLiteral": StringLiteral
 "NumericLiteral": NumericLiteral
 "BigIntLiteral": BigIntLiteral
 "BooleanLiteral": BooleanLiteral
 "NullLiteral": NullLiteral
 "RegExpLiteral": RegExpLiteral
//  "Argument": Argument
 "CallExpression": CallExpression
 "ClassProperty": ClassProperty
 "PrivateProperty": PrivateProperty
 "Param": Param
 "Constructor": Constructor
 "ClassMethod": ClassMethod
 "PrivateMethod": PrivateMethod
 "StaticBlock": StaticBlock
 "Decorator": Decorator
 "FunctionDeclaration": FunctionDeclaration
 "ClassDeclaration": ClassDeclaration
 "VariableDeclaration": VariableDeclaration
 "VariableDeclarator": VariableDeclarator
 "OptionalChainingExpression": OptionalChainingExpression
 "OptionalChainingCall": OptionalChainingCall
 "ThisExpression": ThisExpression
 "ArrayExpression": ArrayExpression
//  "ExprOrSpread": ExprOrSpread
 "ObjectExpression": ObjectExpression
 "SpreadElement": SpreadElement
 "UnaryExpression": UnaryExpression
 "UpdateExpression": UpdateExpression
 "BinaryExpression": BinaryExpression
 "FunctionExpression": FunctionExpression
 "ClassExpression": ClassExpression
 "AssignmentExpression": AssignmentExpression
 "MemberExpression": MemberExpression
 "SuperPropExpression": SuperPropExpression
 "ConditionalExpression": ConditionalExpression
 "Super": Super
 "Import": Import
 "NewExpression": NewExpression
 "SequenceExpression": SequenceExpression
 "ArrowFunctionExpression": ArrowFunctionExpression
 "YieldExpression": YieldExpression
 "MetaProperty": MetaProperty
 "AwaitExpression": AwaitExpression
 "TemplateLiteral": TemplateLiteral
 "TaggedTemplateExpression": TaggedTemplateExpression
 "TemplateElement": TemplateElement
 "ParenthesisExpression": ParenthesisExpression
 "PrivateName": PrivateName
 "JSXMemberExpression": JSXMemberExpression
 "JSXNamespacedName": JSXNamespacedName
 "JSXEmptyExpression": JSXEmptyExpression
 "JSXExpressionContainer": JSXExpressionContainer
 "JSXSpreadChild": JSXSpreadChild
 "JSXOpeningElement": JSXOpeningElement
 "JSXClosingElement": JSXClosingElement
 "JSXAttribute": JSXAttribute
 "JSXText": JSXText
 "JSXElement": JSXElement
 "JSXFragment": JSXFragment
 "JSXOpeningFragment": JSXOpeningFragment
 "JSXClosingFragment": JSXClosingFragment
 "ExportDefaultExpression": ExportDefaultExpression
 "ExportDeclaration": ExportDeclaration
 "ImportDeclaration": ImportDeclaration
 "ExportAllDeclaration": ExportAllDeclaration
 "ExportNamedDeclaration": ExportNamedDeclaration
 "ExportDefaultDeclaration": ExportDefaultDeclaration
 "ImportDefaultSpecifier": ImportDefaultSpecifier
 "ImportNamespaceSpecifier": ImportNamespaceSpecifier
 "NamedImportSpecifier": NamedImportSpecifier
 "ExportNamespaceSpecifier": ExportNamespaceSpecifier
 "ExportDefaultSpecifier": ExportDefaultSpecifier
 "NamedExportSpecifier": NamedExportSpecifier
 "Module": Module
 "Script": Script
 "ArrayPattern": ArrayPattern
 "ObjectPattern": ObjectPattern
 "AssignmentPattern": AssignmentPattern
 "RestElement": RestElement
 "KeyValuePatternProperty": KeyValuePatternProperty
 "AssignmentPatternProperty": AssignmentPatternProperty
 "KeyValueProperty": KeyValueProperty
 "AssignmentProperty": AssignmentProperty
 "GetterProperty": GetterProperty
 "SetterProperty": SetterProperty
 "MethodProperty": MethodProperty
 "ComputedPropName": ComputedPropName
 "BlockStatement": BlockStatement
 "ExpressionStatement": ExpressionStatement
 "EmptyStatement": EmptyStatement
 "DebuggerStatement": DebuggerStatement
 "WithStatement": WithStatement
 "ReturnStatement": ReturnStatement
 "LabeledStatement": LabeledStatement
 "BreakStatement": BreakStatement
 "ContinueStatement": ContinueStatement
 "IfStatement": IfStatement
 "SwitchStatement": SwitchStatement
 "ThrowStatement": ThrowStatement
 "TryStatement": TryStatement
 "WhileStatement": WhileStatement
 "DoWhileStatement": DoWhileStatement
 "ForStatement": ForStatement
 "ForInStatement": ForInStatement
 "ForOfStatement": ForOfStatement
 "SwitchCase": SwitchCase
 "CatchClause": CatchClause
 "TsTypeAnnotation": TsTypeAnnotation
 "TsTypeParameterDeclaration": TsTypeParameterDeclaration
 "TsTypeParameter": TsTypeParameter
 "TsTypeParameterInstantiation": TsTypeParameterInstantiation
 "TsParameterProperty": TsParameterProperty
 "TsQualifiedName": TsQualifiedName
 "TsCallSignatureDeclaration": TsCallSignatureDeclaration
 "TsConstructSignatureDeclaration": TsConstructSignatureDeclaration
 "TsPropertySignature": TsPropertySignature
 "TsGetterSignature": TsGetterSignature
 "TsSetterSignature": TsSetterSignature
 "TsMethodSignature": TsMethodSignature
 "TsIndexSignature": TsIndexSignature
 "TsKeywordType": TsKeywordType
 "TsThisType": TsThisType
 "TsFunctionType": TsFunctionType
 "TsConstructorType": TsConstructorType
 "TsTypeReference": TsTypeReference
 "TsTypePredicate": TsTypePredicate
 "TsImportType": TsImportType
 "TsTypeQuery": TsTypeQuery
 "TsTypeLiteral": TsTypeLiteral
 "TsArrayType": TsArrayType
 "TsTupleType": TsTupleType
 "TsTupleElement": TsTupleElement
 "TsOptionalType": TsOptionalType
 "TsRestType": TsRestType
 "TsUnionType": TsUnionType
 "TsIntersectionType": TsIntersectionType
 "TsConditionalType": TsConditionalType
 "TsInferType": TsInferType
 "TsParenthesizedType": TsParenthesizedType
 "TsTypeOperator": TsTypeOperator
 "TsIndexedAccessType": TsIndexedAccessType
 "TsMappedType": TsMappedType
 "TsLiteralType": TsLiteralType
 "TsTemplateLiteralType": TsTemplateLiteralType
 "TsInterfaceDeclaration": TsInterfaceDeclaration
 "TsInterfaceBody": TsInterfaceBody
 "TsExpressionWithTypeArguments": TsExpressionWithTypeArguments
 "TsTypeAliasDeclaration": TsTypeAliasDeclaration
 "TsEnumDeclaration": TsEnumDeclaration
 "TsEnumMember": TsEnumMember
 "TsModuleDeclaration": TsModuleDeclaration
 "TsModuleBlock": TsModuleBlock
 "TsNamespaceDeclaration": TsNamespaceDeclaration
 "TsImportEqualsDeclaration": TsImportEqualsDeclaration
 "TsExternalModuleReference": TsExternalModuleReference
 "TsExportAssignment": TsExportAssignment
 "TsNamespaceExportDeclaration": TsNamespaceExportDeclaration
 "TsAsExpression": TsAsExpression
//  "TsSatisfiesExpression": TsSatisfiesExpression
 "TsInstantiation": TsInstantiation
 "TsTypeAssertion": TsTypeAssertion
 "TsConstAssertion": TsConstAssertion
 "TsNonNullExpression": TsNonNullExpression
 "Invalid": Invalid
}
