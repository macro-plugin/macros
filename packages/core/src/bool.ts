import { Handler, LabeledMacro, Statement } from "./types";

/**
 * Create a macro plugin that converts `labeled: {}` or `labeled: (...args) => {}` to `$specifier((...args) => {})`,
 * and also `import { $specifier } from $source`
 * @param specifier - the function name
 * @param source - the module path
 * @param allowParams - allow convert the input array to params, default is false.
 * @returns - A labeled macro plugin
 */
export const createLabeledBool: ((id: string) => LabeledMacro) = (id) => {
  return (ast: Statement, code: string, handler: Handler) => {
    // @ts-ignore
    if (ast.type === 'ExpressionStatement' && ast.expression.type == 'BooleanLiteral') {
      // @ts-ignore
      handler.set(id, ast.expression.value);
      return { type: "EmptyStatement" }
    } else {
      throw new Error('this macro only accept a Boolean Literal')
    }
  }
}
