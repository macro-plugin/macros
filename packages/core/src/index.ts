export { createLabeledBlock } from "./block"
export { createLabeledExpr } from "./expr"
export { createSwcPlugin, transform, transformAst, transformAsync } from "./transform"
export { print, printAsync, printExpr, printExprAsync } from "./print"
export { parse, parseAsync, parseExpr, parseType } from "./parse"
export { walk, Walker } from "./walk"
export { macro, $Macro, $LitMacro, $ExprMacro, $TmplMacro, $TypeMacro, $LabeledMacro } from "./macro"
export * from "./api"
export * from "./lib"
export * from "./types"
export * from "./utils"
