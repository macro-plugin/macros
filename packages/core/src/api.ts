import { ArrowFunctionExpression, Expression, FunctionDeclaration, FunctionExpression, Invalid, Param, TsType } from "@swc/core"
import type { BaseNode, ExprMacro, LabeledMacro, MacroPlugin, WalkContext } from "./types"

import { isRegExp } from "./utils"
import { parseExpr } from "./parse"
import { walk } from "./walk"

export function createMacro (macro: MacroPlugin) {
  return macro
}

function createLit (this: WalkContext, value: unknown): BaseNode {
  if (value === undefined) {
    return {
      type: "Identifier",
      span: {
        start: 0,
        end: 0,
        ctxt: 2
      },
      value: "undefined",
      optional: false
    }
  }
  if (value && typeof value === "object") {
    if (isRegExp(value)) return this.parseExpr(value.toString())
    if ("span" in value && "type" in value) return value as BaseNode
  }
  if (typeof value === "function") return this.parseExpr(value.toString())
  return this.parseExpr(JSON.stringify(value))
}

export function createLitMacro(map: Record<string, unknown>): MacroPlugin;
export function createLitMacro<T>(key: string, value: T): MacroPlugin;
export function createLitMacro (arg: string | Record<string, unknown>, value?: unknown): MacroPlugin {
  return createMacro(typeof arg === "string"
    ? {
      Identifier (ast) {
        if (ast.value === arg && !this.track(arg)) return createLit.apply(this, [value])
      }
    }
    : {
      Identifier (ast) {
        if (ast.value in arg && !this.track(ast.value)) return createLit.apply(this, [arg[ast.value]])
      }
    })
}

function flatExpr (f: Function, args: Expression[], typeParams?: TsType[], optional = false): Expression {
  const ast = parseExpr(f.toString()) as FunctionDeclaration | FunctionExpression | ArrowFunctionExpression
  const params: Record<string, { value?: Expression }> = {}

  ast.params.forEach((p, i) => {
    const pat = ast.type === "ArrowFunctionExpression" ? p : ((p as Param).pat)
    if (pat.type === "Identifier") {
      params[pat.value] = args[i] ? { value: args[i] } : {}
    } else if (pat.type === "AssignmentPattern") {
      if (pat.left.type === "Identifier") {
        params[pat.left.value] = { value: args[i] || pat.right }
      }
    }
  })

  let output: Invalid | Expression = {
    type: "Invalid",
    span: {
      start: 0,
      end: 0,
      ctxt: 0
    }
  }

  if (ast.body) {
    walk(ast.body, {
      // @ts-ignore
      enter (ast: BaseNode) {
        if (ast.type === "Identifier" && ast.value in params) {
          const v = params[ast.value].value
          if (v) this.replace(v)
        }
      },
      // @ts-ignore
      leave (ast: BaseNode) {
        if (ast.type === "ReturnStatement" && ast.argument) {
          output = ast.argument
        }
      }
    })

    if (ast.body.type !== "BlockStatement") output = ast.body
  }

  return output
}

export function createExprMacro (name: string, f: Function | ExprMacro | { enter?: ExprMacro, leave?: ExprMacro }): MacroPlugin {
  if (typeof f === "object") {
    return createMacro({
      CallExpression: {
        enter (ast) {
          if (f.enter && ast.callee.type === "Identifier" && ast.callee.value === name && !this.track(ast.callee.value)) {
            return f.enter.apply(this, [ast.arguments.map(i => i.expression), ast.typeArguments?.params, ast.callee.optional])
          }
        },
        leave (ast) {
          if (f.leave && ast.callee.type === "Identifier" && ast.callee.value === name && !this.track(ast.callee.value)) {
            return f.leave.apply(this, [ast.arguments.map(i => i.expression), ast.typeArguments?.params, ast.callee.optional])
          }
        }
      }
    })
  }
  return createMacro({
    CallExpression (ast) {
      if (ast.callee.type === "Identifier" && ast.callee.value === name && !this.track(ast.callee.value)) {
        const args = ast.arguments.map(i => i.expression)
        const tys = ast.typeArguments?.params
        return f.toString().startsWith("function") ? (f as ExprMacro).apply(this, [args, tys, ast.callee.optional]) : flatExpr(f, args, tys, ast.callee.optional)
      }
    }
  })
}

export function createTypeMacro () {

}

export function createTmplMacro (tag: string,) {

}

export function createLabeledMacro (label: string, f: LabeledMacro | {
  enter?: LabeledMacro,
  leave?: LabeledMacro,
}) {
  return createMacro({
    LabeledStatement: {
      enter (ast, parent, prop, index) {
        if (ast.label.value === label) return (typeof f === "object" ? f.enter : f)?.apply(this, [ast.body, parent, prop, index])
      },
      leave (ast, parent, prop, index) {
        if (ast.label.value === label && typeof f === "object") return f.leave?.apply(this, [ast.body, parent, prop, index])
      }
    }
  })
}
