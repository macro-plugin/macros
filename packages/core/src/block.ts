import { ArrowFunctionExpression, ExpressionStatement } from "@swc/core";

import { MacroPlugin } from "./types";
import { createMacro } from "./api";

/**
 * Create a macro plugin that converts `labeled: {}` or `labeled: (...args) => {}` to `$specifier((...args) => {})`,
 * and also `import { $specifier } from $source`
 * @param specifier - the function name
 * @param source - the module path
 * @param allowParams - allow convert the input array to params, default is false.
 * @returns - A labeled macro plugin
 */
export const createLabeledBlock: ((label: string, specifier: string, source: string, allowParams?: boolean) => MacroPlugin) = (label, specifier, source, allowParams = false) => {
  return createMacro({
    LabeledStatement(ast) {
      if (ast.label.value != label) return;
      this.import([{ name: specifier }], source);
      if (ast.body.type == "BlockStatement") {
        return ({
          type: "ExpressionStatement",
          expression: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              value: specifier,
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            arguments: [
              {
                "expression": {
                  type: 'ArrowFunctionExpression',
                  generator: false,
                  async: false,
                  params: [],
                  body: ast.body,
                  span: {
                    start: 0,
                    end: 0,
                    ctxt: 0
                  }
                } as ArrowFunctionExpression
              }
            ],
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          },
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        }) as ExpressionStatement
      } else if (ast.body.type == "ExpressionStatement" && ast.body.expression.type == 'ArrowFunctionExpression') {
        return ({
          type: "ExpressionStatement",
          expression: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              value: specifier,
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            arguments: [
              {
                expression: ast.body.expression
              }
            ],
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          },
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        }) as ExpressionStatement
      } else if (allowParams && ast.body.type == "ExpressionStatement" && ast.body.expression.type == 'ArrayExpression') {
        return ({
          type: "ExpressionStatement",
          expression: {
            type: 'CallExpression',
            callee: {
              type: 'Identifier',
              value: specifier,
              span: {
                start: 0,
                end: 0,
                ctxt: 0
              }
            },
            arguments: ast.body.expression.elements,
            span: {
              start: 0,
              end: 0,
              ctxt: 0
            }
          },
          span: {
            start: 0,
            end: 0,
            ctxt: 0
          }
        }) as ExpressionStatement
      } else {
        throw new Error('this macro only accept a ArrowFunction or BlockStatement' + (allowParams ? ' or an ArrayExpression' : ''))
      }
    }
  })
}
