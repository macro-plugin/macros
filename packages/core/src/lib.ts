import { Expression, Identifier } from "@swc/core"
import { createExprMacro, createTmplMacro } from "./api"
import { evalAst, evalExpr, span } from "./utils"

import { printExpr } from "./print"
import { walk } from "./walk"

export var $Eval = createExprMacro("$Eval", function (args) {
  const t = args[0].type

  if (t === "StringLiteral") return this.parseExpr(evalExpr(args[0].value))
  if (t === "TemplateLiteral") {
    const track = (name: string) => this.track(name)
    walk(args[0], {
      enter (node) {
        if (node.type === "Identifier") {
          const v = track((node as Identifier).value)
          if (v && v.value) this.replace(v.value)
        }
      }
    })

    return this.parseExpr(evalExpr(evalExpr(this.printExpr(args[0]))))
  }

  if (t === "FunctionExpression" || t === "ArrowFunctionExpression") return this.parseExpr(evalExpr(`(${this.printExpr(args[0])})(${args.slice(1).map(i => this.printExpr(i)).join(",")})`))
  return this.parseExpr(evalAst(args[0]).toString())
}, "(<T>(expr: string) => T) & (<T>(expr: T) => T) & (<F extends (...args: any) => any>(expr: F, ...args: Parameters<F>) => ReturnType<F>)")

export const printAst = (ast: object) => JSON.stringify(ast).replace(/("(start|end|ctxt)":\s*)(\d+)/g, "$10")

export var $Ast = createExprMacro("$Ast", function (args) {
  return this.parseExpr(printAst(args[0].type === "StringLiteral" ? this.parseExpr(args[0].value) : args[0]))
}, '<T>(expr: T) => import("@swc/core").Expression')

export var $Env = createExprMacro("$Env", function (args, typeArgs) {
  if (args[0].type !== "StringLiteral") throw new Error("$Env macro only support StringLiteral as input.")
  const value = process.env[args[0].value] ?? ""

  if (typeArgs && typeArgs?.[0].type === "TsKeywordType") {
    switch (typeArgs[0].kind) {
      case "boolean":
        return {
          type: "BooleanLiteral",
          value: !!value,
          span
        }
      case "number":
        return {
          type: "NumericLiteral",
          value: +value,
          span
        }
    }
  }

  return {
    type: "StringLiteral",
    span,
    value
  }
}, "<R>(key: string) => R")

export var $Stringify = createExprMacro("$Stringify", function (args, typeArgs) {
  return {
    type: "StringLiteral",
    span,
    value: typeArgs && typeArgs.length > 0 ? this.printType(typeArgs[0]) : this.printExpr(args[0])
  }
}, "((expr: any) => string) & (<T>() => string)")

export const printTmpl = (strings: string[], exprs: Expression[]) => strings.reduce((query, queryPart, i) => {
  const valueExists = i < exprs.length
  const text = query + queryPart

  return valueExists ? text + printExpr(exprs[i]).code : text
}, "")

export const printRawTmpl = (strings: string[], exprs: unknown[]) => strings.reduce((query, queryPart, i) => {
  const valueExists = i < exprs.length
  const text = query + queryPart

  return valueExists ? text + exprs[i] : text
}, "")

function walkObject (obj: object | null, onEnter: (k: string, v: unknown, parent: Record<string, any>) => unknown, grand?: Record<string, any>, grandKey?: string) {
  if (obj && typeof obj === "object") {
    const allKeys = Object.keys(obj)
    for (let i = 0; i < allKeys.length; i++) {
      const k = allKeys[i]
      const v = obj[k as keyof typeof obj]

      const r = onEnter(k, v, obj)
      if (r != null && grand) {
        grand[grandKey!] = r
        break
      }

      if (typeof v === "object") {
        walkObject(v, onEnter, obj, k)
      }
    }
  }
  return obj
}

export var $Expr = createTmplMacro("$Expr", function (strings, ...exprs) {
  const exprMarkers: string[] = exprs.map((_, i) => "_macro_marker_" + i + "_")
  const options = { syntax: "typescript", tsx: true } as const
  const ast = this.parseExpr(printRawTmpl(strings, exprMarkers), options)

  walkObject(ast, (k, v, parent) => {
    if (k === "span") {
      parent[k] = span
      return
    }

    if (typeof v === "string") {
      const i = exprMarkers.findIndex((m) => m === v)
      if (i !== -1) return "__macro$$Start__" + this.printExpr(exprs[i]) + "__macro$$End__"
    }
  })

  const expr = JSON.stringify(ast, undefined, 2).replace(/("__macro\$\$Start__)|(__macro\$\$End__")/g, "").replace(/\\"/g, '"')

  return this.parseExpr(expr)
}, "import(\"@swc/core\").Expression")

export var $Quote = createTmplMacro("$Quote", function (strings, ...exprs) {
  const exprMarkers: string[] = exprs.map((_, i) => "_macro_marker_" + i + "_")
  const options = { syntax: "typescript", tsx: true } as const
  const ast = this.parse(printRawTmpl(strings, exprMarkers), options)

  walkObject(ast, (k, v, parent) => {
    if (k === "span") {
      parent[k] = span
      return
    }

    if (typeof v === "string") {
      const i = exprMarkers.findIndex((m) => m === v)
      if (i !== -1) return "__macro$$Start__" + this.printExpr(exprs[i]) + "__macro$$End__"
    }
  })

  const expr = JSON.stringify(ast, undefined, 2).replace(/("__macro\$\$Start__)|(__macro\$\$End__")/g, "").replace(/\\"/g, '"')

  return this.parseExpr(expr)
}, "(import(\"@swc/core\").Expression)[]")
