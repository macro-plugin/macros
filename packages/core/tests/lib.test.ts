import { $Ast, $Column, $Concat, $Env, $Eval, $Expr, $ID, $Include, $IncludeJSON, $Line, $Quote, $Span, $Stringify, transform } from "../src"

test("$Eval macro", () => {
  expect(transform("$Eval(1 + 2 / 10)", { macros: [$Eval] }).code).toEqual("1.2;\n")
  expect(transform("$Eval('1 + 3')", { macros: [$Eval] }).code).toEqual("4;\n")
  // eslint-disable-next-line no-template-curly-in-string
  expect(transform("let a = 3; let b = 4; $Eval(`${a} + ${b}`)", { macros: [$Eval] }).code).toEqual("let a = 3;\nlet b = 4;\n7;\n")
  expect(transform("$Eval((a, b) => { return a + b }, 1, 2)", { macros: [$Eval] }).code).toEqual("3;\n")
})

test("$Ast macro", () => {
  expect(transform(`
  const i = $Ast(abc)
  const d = $Ast('abc')
  const a = $Ast("'hello'")
  const b = $Ast(1 + 2)
  const c = $Ast('a * 3')
  `, { macros: [$Ast] }).code).toMatchSnapshot()
})

test("$Quote macro", () => {
  const r = transform(`
  const someNumber = $Ast(5)
  const someIdent = $Ast('hello')
  const someExpr = $Expr\`3 + \${someNumber}\`
  const tokens = $Quote\`
    let \${someIdent} = \${someExpr}
  \`
  `, { macros: [$Ast, $Expr, $Quote], emitDts: true, jsc: { parser: { syntax: "typescript", tsx: true } } })
  expect(r.dts).toMatchSnapshot()
  expect(r.code).toMatchSnapshot()
})

test("$Stringify macro", () => {
  expect(transform(`
  const a = $Stringify(abc)
  const b = $Stringify('abc')
  const c = $Stringify(1 + 2)
  const d = $Stringify(hello(1, 2))
  const e = $Stringify<number | string>()
  `, { macros: [$Stringify], jsc: { parser: { syntax: "typescript" } } }).code).toMatchSnapshot()
})

test("$Env macro", () => {
  expect(transform(`
  const e = $Env("NODE_ENV")
  const s = $Env<string>("NODE_ENV")
  const b = $Env<boolean>("DEV")
  const n = $Env<number>("DEV")
  `, { macros: [$Env], jsc: { parser: { syntax: "typescript" } } }).code).toMatchSnapshot()
})

test("$Span macro", () => {
  const code = "$Span()"
  expect(code.slice(0, 7)).toBe("$Span()")
  expect(/^\s*\[\s*0,\s*7/.test(transform(code, { macros: [$Span] }).code)).toBe(true)

  const multiline = `
    const a = $Span()
  `
  expect(multiline.slice(15, 22)).toBe("$Span()")
  expect(/\s*\[\s*15,\s*22/.test(transform(multiline, { macros: [$Span] }).code)).toBe(true)
})

test("$Line macro", () => {
  expect(transform(`const ln = $Line()
  const ln2 = $Line()
  const ln3 = $Line()
  `, { macros: [$Line] }).code).toEqual("const ln = 1;\nconst ln2 = 2;\nconst ln3 = 3;\n")
})

test("$Column macro", () => {
  expect(transform(`$Column()
  $Column()
  const col = $Column()
  $Column()`, { macros: [$Column] }).code).toEqual("1;\n3;\nconst col = 15;\n3;\n")
})

test("$ID macro", () => {
  expect(transform("const id = $ID()\nconst id2 = $ID()", { macros: [$ID] }).code).toEqual("const id = \"lb6ods\";\nconst id2 = \"vzqptz\";\n")
})

test("$Include macro", () => {
  expect(transform("$Include('packages/factory/src/index.ts')", { macros: [$Include] }).code).toMatchSnapshot()
})

test("$IncludeJSON macro", () => {
  expect(transform("$IncludeJSON('packages/core/package.json')", { macros: [$IncludeJSON] }).code.includes("@macro-plugin/core")).toBe(true)
  expect(transform("$IncludeJSON('packages/core/package.json', 'name')", { macros: [$IncludeJSON] }).code).toBe("\"@macro-plugin/core\";\n")
  expect(transform("$IncludeJSON('package.json', 'private')", { macros: [$IncludeJSON] }).code).toBe("true;\n")
})

test("$Concat macro", () => {
  expect(transform("$Concat('hello', 'world', 1, 2, true, undefined, null, 3.14)", { macros: [$Concat] }).code).toEqual("\"helloworld12trueundefinednull3.14\";\n")
  expect(() => transform("$Concat('hello', abc)", { macros: [$Concat] }).code).toThrow()
})
