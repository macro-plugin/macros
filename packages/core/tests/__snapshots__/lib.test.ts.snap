// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$Ast macro 1`] = `
"const i = {
    "type": "Identifier",
    "span": {
        "start": 0,
        "end": 0,
        "ctxt": 0
    },
    "value": "abc",
    "optional": false
};
const d = {
    "type": "Identifier",
    "span": {
        "start": 0,
        "end": 0,
        "ctxt": 0
    },
    "value": "abc",
    "optional": false
};
const a = {
    "type": "StringLiteral",
    "span": {
        "start": 0,
        "end": 0,
        "ctxt": 0
    },
    "value": "hello",
    "raw": "'hello'"
};
const b = {
    "type": "BinaryExpression",
    "span": {
        "start": 0,
        "end": 0,
        "ctxt": 0
    },
    "operator": "+",
    "left": {
        "type": "NumericLiteral",
        "span": {
            "start": 0,
            "end": 0,
            "ctxt": 0
        },
        "value": 1,
        "raw": "1"
    },
    "right": {
        "type": "NumericLiteral",
        "span": {
            "start": 0,
            "end": 0,
            "ctxt": 0
        },
        "value": 2,
        "raw": "2"
    }
};
const c = {
    "type": "BinaryExpression",
    "span": {
        "start": 0,
        "end": 0,
        "ctxt": 0
    },
    "operator": "*",
    "left": {
        "type": "Identifier",
        "span": {
            "start": 0,
            "end": 0,
            "ctxt": 0
        },
        "value": "a",
        "optional": false
    },
    "right": {
        "type": "NumericLiteral",
        "span": {
            "start": 0,
            "end": 0,
            "ctxt": 0
        },
        "value": 3,
        "raw": "3"
    }
};
"
`;

exports[`$Env macro 1`] = `
"const e = "test";
const s = "test";
const b = false;
const n = 0;
"
`;

exports[`$Include macro 1`] = `
"(()=>{
    "use strict";
    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    _exportStar(require("./runtime"), exports);
    _exportStar(require("./macros"), exports);
    function _exportStar(from, to) {
        Object.keys(from).forEach(function(k) {
            if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        });
        return from;
    }
})();
"
`;

exports[`$Quote macro 1`] = `
"declare global {
    const $Ast: <T>(expr: T) => import("@swc/core").Expression;
    const $Expr: (strings: TemplateStringsArray, ...expressions: unknown[]) => import("@swc/core").Expression;
    const $Quote: (strings: TemplateStringsArray, ...expressions: unknown[]) => (import("@swc/core").Expression)[];
}
export {}
"
`;

exports[`$Quote macro 2`] = `
"const someNumber = {
    "type": "NumericLiteral",
    "span": {
        "start": 0,
        "end": 0,
        "ctxt": 0
    },
    "value": 5,
    "raw": "5"
};
const someIdent = {
    "type": "Identifier",
    "span": {
        "start": 0,
        "end": 0,
        "ctxt": 0
    },
    "value": "hello",
    "optional": false
};
const someExpr = {
    "type": "BinaryExpression",
    "span": {
        "start": 0,
        "end": 0,
        "ctxt": 0
    },
    "operator": "+",
    "left": {
        "type": "NumericLiteral",
        "span": {
            "start": 0,
            "end": 0,
            "ctxt": 0
        },
        "value": 3,
        "raw": "3"
    },
    "right": someNumber
};
const tokens = [
    {
        "type": "VariableDeclaration",
        "span": {
            "start": 0,
            "end": 0,
            "ctxt": 0
        },
        "kind": "let",
        "declare": false,
        "declarations": [
            {
                "type": "VariableDeclarator",
                "span": {
                    "start": 0,
                    "end": 0,
                    "ctxt": 0
                },
                "id": someIdent,
                "init": someExpr,
                "definite": false
            }
        ]
    }
];
"
`;

exports[`$Stringify macro 1`] = `
"const a = "abc";
const b = '"abc"';
const c = "1 + 2";
const d = "hello(1, 2)";
const e = "number | string";
"
`;
